File: hellojava/src/test/java/com/fathzer/chess/BoardTest.java
package com.fathzer.chess;

import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

import static com.fathzer.chess.MoveGeneratorsTest.*;

import java.util.LinkedList;
import java.util.List;
import java.util.function.IntConsumer;
import java.util.stream.IntStream;

class BoardTest {

    @Test
    void testFenParsing() {
        // Standard chess starting position FEN (only piece placement part)
        String fen = "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR";
        final Board board = new Board(fen);
        assertNotNull(board);
        assertEquals(Piece.WHITE_QUEEN, board.getPiece("d1"));

        // Test that blockers are set correctly
        IntConsumer expectBlocker = i -> assertEquals(Piece.BLOCKER, board.getPiece(i), "expected blocker at " + i);
        IntStream.range(0, 20).forEach(expectBlocker);
        IntStream.range(100, 120).forEach(expectBlocker);
    }

    @Test
    void testCopyConstructor() {
        String fen = "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR";
        Board original = new Board(fen);
        Board copy = new Board(original);
        assertNotSame(original, copy);
    }

    @Test
    void testInvalidFenThrows() {
        String badFileCount = "rnbqkbnr/pppppppp/9/8/8/8/PPPPPPPP/RNBQKBNR"; // 9 is invalid
        assertThrows(IllegalArgumentException.class, () -> new Board(badFileCount));
        String badFileCount2 = "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBN"; // Last row missed one file
        assertThrows(IllegalArgumentException.class, () -> new Board(badFileCount2));
        String badFileCount3 = "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPP2/RNBQKBNR"; // 7th row has an extra file
        assertThrows(IllegalArgumentException.class, () -> new Board(badFileCount3));

        String badRankCount = "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR/"; // Last rank is invalid
        assertThrows(IllegalArgumentException.class, () -> new Board(badRankCount));
        String badRankCount2 = "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP"; // Last rank is missing
        assertThrows(IllegalArgumentException.class, () -> new Board(badRankCount2));

        String badPiece = "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQXBNR"; // X is not a valid piece
        assertThrows(IllegalArgumentException.class, () -> new Board(badPiece));
    }
    
    @Test
    void testMakeMove() {
        // Start with initial position
        Board board = new Board("rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR");

        // Make a pawn move e2-e4
        Move move = new Move("e2", "e4"); // e2-e4
        board.makeMove(move);
        
        // Verify the move was made
        assertNull(board.getPiece("e2"), "Source square should be empty after move");
        assertEquals(Piece.WHITE_PAWN, board.getPiece("e4"), "Piece should be at destination");
    }
    
    @Test
    void testMakeCapture() {
        // Set up a position where white can capture a pawn
        Board board = new Board("rnbqkbnr/ppp1pppp/8/3p4/4P3/8/PPPP1PPP/RNBQKBNR");
        
        // Make a capture exd5
        Move capture = new Move("e4", "d5"); // e4xd5
        board.makeMove(capture);
        
        // Verify the capture
        assertNull(board.getPiece("e4"), "Source square should be empty after capture");
        assertEquals(Piece.WHITE_PAWN, board.getPiece("d5"), "Capturing piece should be at destination");
    }
    
    @Test
    void testUnmakeMove() {
        // Start with initial position
        Board original = new Board("rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR");
        Board board = new Board(original);
        
        // Make a move
        Move move = new Move("e2", "e4"); // e2-e4
        board.makeMove(move);
        
        // Unmake the move
        board.unmakeMove();
        
        // Board should be back to original state
        for (char rank = '1'; rank < '8'; rank++) {
            for (char file = 'a'; file < 'h'; file++) {
                assertEquals(original.getPiece(file + "" + rank), board.getPiece(file + "" + rank),
                    String.format("Pieces should match at %c%c", file, rank));
            }
        }
    }
    
    @Test
    void testUnmakeCapture() {
        // Set up a position where white can capture a pawn
        Board original = new Board("rnbqkbnr/ppp1pppp/8/3p4/4P3/8/PPPP1PPP/RNBQKBNR");
        Board board = new Board(original);
        
        // Make a capture
        Move capture = new Move("e4", "d5"); // e4xd5
        board.makeMove(capture);
        
        // Unmake the capture
        board.unmakeMove();
        
        // Board should be back to original state
        for (char rank = '1'; rank < '8'; rank++) {
            for (char file = 'a'; file < 'h'; file++) {
                assertEquals(original.getPiece(file + "" + rank), board.getPiece(file + "" + rank),
                    String.format("Pieces should match at %c%c", file, rank));
            }
        }
    }
    
    @Test
    void testUnmakeWithoutMoves() {
        Board board = new Board("rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR");
        assertThrows(IllegalStateException.class, board::unmakeMove,
            "Should throw when no moves to unmake"
        );
    }

    @Test
    void testGetMoves() {
        final Board board = new Board("8/8/8/8/1k6/8/pK6/Q7");
        final List<Move> whiteMoves = board.getMoves(true);
        List<Move> expected = new LinkedList<>();
        expected.addAll(parseMoveList("a1", "a2 b1 c1 d1 e1 f1 g1 h1"));
        expected.addAll(parseMoveList("b2", "a2 a3 b3 b3 c2 c1 b1"));
        testMoves(expected, whiteMoves);
        final List<Move> blackMoves = board.getMoves(false);
        testMoves(parseMoveList("b4", "a5 b5 c5 a4 c4 a3 b3 c3"), blackMoves);
        assertEquals(8, blackMoves.size());
    }
}

File: hellogo/pkg/chess/board_test.go
package chess

import (
	"testing"
)

const (
	startingFEN = "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR"
)

func TestFenParsing(t *testing.T) {
	// Standard chess starting position FEN (only piece placement part)
	board, err := NewBoard(startingFEN)
	if err != nil {
		t.Fatalf("Failed to parse FEN: %v", err)
	}

	// Test specific piece positions
	piece := board.GetPiece("d1")
	if piece == nil || !piece.Equals(&WHITE_QUEEN) {
		t.Errorf("Expected white queen at d1, got %v", piece)
	}

	// Test that blockers are set correctly
	testBlocker := func(i int) {
		if board.pieces[i] != &BLOCKER {
			t.Errorf("Expected blocker at index %d, got %v", i, board.pieces[i])
		}
	}

	// Test border blockers
	for i := 0; i < 20; i++ {
		testBlocker(i)
	}
	for i := 100; i < 120; i++ {
		testBlocker(i)
	}
}

func TestCopyConstructor(t *testing.T) {
	original, err := NewBoard(startingFEN)
	if err != nil {
		t.Fatalf("Failed to create board: %v", err)
	}

	copy := NewBoardCopy(original)

	// Verify it's a deep copy
	if copy == original {
		t.Error("Copy should be a different instance")
	}

	// Verify the pieces slice is a different instance
	if &copy.pieces[0] == &original.pieces[0] {
		t.Error("Pieces slice should be a deep copy")
	}

	// Make a move on the copy and verify original is unchanged
	move, _ := NewMoveFromStrings("e2", "e4")
	copy.MakeMove(&move)

	if copy.GetPiece("e4") == nil || copy.GetPiece("e2") != nil {
		t.Error("Move should be applied to copy")
	}

	if original.GetPiece("e4") != nil || !original.GetPiece("e2").Equals(&WHITE_PAWN) {
		t.Error("Original board should be unchanged")
	}
}

func TestInvalidFenThrows(t *testing.T) {
	tests := []struct {
		name string
		fen  string
	}{
		{"badFileCount1", "rnbqkbnr/pppppppp/9/8/8/8/PPPPPPPP/RNBQKBNR"},  // 9 is invalid
		{"badFileCount2", "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBN"},   // Last row missed one file
		{"badFileCount3", "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPP2/RNBQKBNR"},  // 7th row has an extra file
		{"badRankCount1", "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR/"}, // Extra rank
		{"badRankCount2", "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP"},           // Missing rank
		{"badPiece", "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQXBNR"},       // X is not a valid piece
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			_, err := NewBoard(tt.fen)
			if err == nil {
				t.Errorf("Expected error for FEN: %s", tt.fen)
			}
		})
	}
}

func TestMakeMove(t *testing.T) {
	// Start with initial position
	board, _ := NewBoard(startingFEN)

	// Make a pawn move e2-e4
	move, _ := NewMoveFromStrings("e2", "e4")
	err := board.MakeMove(&move)
	if err != nil {
		t.Fatalf("Failed to make move: %v", err)
	}

	// Verify the move was made
	if board.GetPiece("e2") != nil {
		t.Error("Source square should be empty after move")
	}
	if board.GetPiece("e4") == nil || !board.GetPiece("e4").Equals(&WHITE_PAWN) {
		t.Error("Piece should be at destination")
	}
}

func TestMakeCapture(t *testing.T) {
	// Set up a position where white can capture a pawn
	board, _ := NewBoard("rnbqkbnr/ppp1pppp/8/3p4/4P3/8/PPPP1PPP/RNBQKBNR")

	// Make a capture exd5
	capture, _ := NewMoveFromStrings("e4", "d5")
	err := board.MakeMove(&capture)
	if err != nil {
		t.Fatalf("Failed to make capture: %v", err)
	}

	// Verify the capture
	if board.GetPiece("e4") != nil {
		t.Error("Source square should be empty after capture")
	}
	if board.GetPiece("d5") == nil || !board.GetPiece("d5").Equals(&WHITE_PAWN) {
		t.Error("Capturing piece should be at destination")
	}
}

func TestUnmakeMove(t *testing.T) {
	// Start with initial position
	original, _ := NewBoard(startingFEN)
	board := NewBoardCopy(original)

	// Make a move e2-e4
	move, _ := NewMoveFromStrings("e2", "e4")
	board.MakeMove(&move)

	// Unmake the move
	err := board.UnmakeMove()
	if err != nil {
		t.Fatalf("Failed to unmake move: %v", err)
	}

	// Board should be back to original state
	for rank := '1'; rank <= '8'; rank++ {
		for file := 'a'; file <= 'h'; file++ {
			square := string([]rune{file, rank})
			expected := original.GetPiece(square)
			actual := board.GetPiece(square)
			if expected != actual {
				t.Errorf("Pieces don't match at %s: expected %v, got %v", square, expected, actual)
			}
		}
	}
}

func TestUnmakeCapture(t *testing.T) {
	// Set up a position where white can capture a pawn
	original, _ := NewBoard("rnbqkbnr/ppp1pppp/8/3p4/4P3/8/PPPP1PPP/RNBQKBNR")
	board := NewBoardCopy(original)

	// Make a capture exd5
	capture, _ := NewMoveFromStrings("e4", "d5")
	board.MakeMove(&capture)

	// Unmake the capture
	err := board.UnmakeMove()
	if err != nil {
		t.Fatalf("Failed to unmake capture: %v", err)
	}

	// Board should be back to original state
	for rank := '1'; rank <= '8'; rank++ {
		for file := 'a'; file <= 'h'; file++ {
			square := string([]rune{file, rank})
			expected := original.GetPiece(square)
			actual := board.GetPiece(square)
			if expected != actual {
				t.Errorf("Pieces don't match at %s: expected %v, got %v", square, expected, actual)
			}
		}
	}
}

func TestUnmakeWithoutMoves(t *testing.T) {
	board, _ := NewBoard("rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR")
	err := board.UnmakeMove()
	if err == nil {
		t.Error("Expected error when unmaking move with no moves to unmake")
	}
}

func TestGetMoves(t *testing.T) {
	board, _ := NewBoard("8/8/8/8/1k6/8/pK6/Q7")

	// Test white moves
	whiteMoves := board.GetMoves(true)
	expectedWhiteMoves := []Move{}

	// Add queen moves
	for _, to := range []string{"a2", "b1", "c1", "d1", "e1", "f1", "g1", "h1"} {
		m, _ := NewMoveFromStrings("a1", to)
		expectedWhiteMoves = append(expectedWhiteMoves, m)
	}

	// Add king moves
	for _, to := range []string{"a3", "b3", "c3", "a2", "c2", "b1", "c1"} {
		m, _ := NewMoveFromStrings("b2", to)
		expectedWhiteMoves = append(expectedWhiteMoves, m)
	}

	// Check white moves
	if len(whiteMoves) != len(expectedWhiteMoves) {
		t.Errorf("Expected %d white moves, got %d", len(expectedWhiteMoves), len(whiteMoves))
		// Print actual moves for debugging
		for i, m := range whiteMoves {
			t.Logf("Move %d: %s -> %s", i, getUCI(m.From()), getUCI(m.To()))
		}
	}

	// Verify each expected move exists in actual moves
	for _, expected := range expectedWhiteMoves {
		found := false
		for _, actual := range whiteMoves {
			if actual.From() == expected.From() && actual.To() == expected.To() {
				found = true
				break
			}
		}
		if !found {
			t.Errorf("Expected move %s -> %s not found in actual moves",
				getUCI(expected.From()), getUCI(expected.To()))
		}
	}

	// Test black moves
	blackMoves := board.GetMoves(false)
	expectedBlackSquares := map[string]bool{
		"a5": true, "b5": true, "c5": true,
		"a4": true, "c4": true,
		"a3": true, "b3": true, "c3": true,
	}

	if len(blackMoves) != len(expectedBlackSquares) {
		t.Errorf("Expected %d black moves, got %d", len(expectedBlackSquares), len(blackMoves))
		// Print actual moves for debugging
		for i, m := range blackMoves {
			t.Logf("Black move %d: %s -> %s", i, getUCI(m.From()), getUCI(m.To()))
		}
	}

	// Verify each actual move is in expected squares
	for _, move := range blackMoves {
		toSquare := getUCI(move.To())
		if !expectedBlackSquares[toSquare] {
			t.Errorf("Unexpected black move to %s", toSquare)
		}
	}
}
